ignite/engine/engine.py

```diff
@@ -982,11 +991,14 @@ def _internal_run_as_gen(self) -> Generator[Any, None, State]:
                     # time is available for handlers but must be updated after fire
                     self.state.times[Events.EPOCH_COMPLETED.name] = epoch_time_taken
 
-                    handlers_start_time = time.time()
-                    self._fire_event(Events.EPOCH_COMPLETED)
-                    epoch_time_taken += time.time() - handlers_start_time
-                    # update time wrt handlers
-                    self.state.times[Events.EPOCH_COMPLETED.name] = epoch_time_taken
+                    if self.should_terminate_single_epoch != "skip_epoch_completed":
+                        handlers_start_time = time.time()
+                        self._fire_event(Events.EPOCH_COMPLETED)
+                        epoch_time_taken += time.time() - handlers_start_time
+                        # update time wrt handlers
+                        self.state.times[Events.EPOCH_COMPLETED.name] = epoch_time_taken
+
+                    self.should_terminate_single_epoch = False
```

Comment:
I had an impression that we do not need to reset `self.should_terminate_single_epoch` as it should be reset by the existing code, same as for `True` value. Do you think this really necessary?

Reply:
I tried to comment out line 1001, but now it's failing the tests. I guess the problem is that it was necessary to remove the reset of `should_terminate_single_epoch` from the body of `except _EngineTerminateSingleEpochException`.
The flag can be reset only after the if statement where we check whether the `EPOCH_COMPLETED` should be fired or not...

Reply:
This code:
```python
                    self.should_terminate_single_epoch = False
                    yield from self._maybe_terminate_or_interrupt()
```
however looks a bit suspicious as it seems that `_maybe_terminate_or_interrupt` wont raise `_EngineTerminateSingleEpochException` when user calls `terminate_epoch()`:
```python
        if self.should_terminate_single_epoch:
            raise _EngineTerminateSingleEpochException()
```

Reply:
To my understanding of the code, right now `_EngineTerminateSingleEpochException` is supposed to be handled within `_run_once_on_dataset_as_gen()`. I guess that's why the exception is not caught  in `_internal_run_as_gen()`. That being said, TBH I also find these two functions (and `_maybe_terminate_or_interrupt()`) a bit confusing.

Reply:
`*_as_gen()` functions were introduced together with `_legacy()` to support `engine.interrupt()` use-case where we can interrupt the engine from the handler, execute other code and come back to the engine executing without using handlers:
```python

engine = Engine(lambda e, b: None)

@engine.on(Events.ITERATION_COMPLETED(once=11))
def call_interrupt():
    engine.interrupt()

print("Start engine run with interruptions:")
state = engine.run(data, max_epochs=max_epochs)

print("1 Engine run is interrupted at ", state.epoch, state.iteration)
# Here we can now execute any other code
a, b = 1, 2
c = a + b

# Continue the run
state = engine.run(data, max_epochs=max_epochs)
# ...
```